proposals/0000-local-modules.rst

```diff
@@ -194,143 +195,294 @@ Motivation
+  symbolic names can be qualified.
+  
+* **Entity**: An entity is a definition that can be exported and imported.
+  An entity has a name. Entities include variables, classes, datatypes, and
+  constructors, among a few other constructs.
+
+* **Environment**: An environment is a set of names and qualified names,
+  used for validating name occurrences.
+
+* **in scope**: A name is said to be **in scope** when it is in the current
+  environment.
+  
+* **Scope**: A scope is a lexical region of a program where all name occurrences are
+  looked up in an environment (or a superset of an environment).
+
+* **Module**: A module is a set of declarations of entities, along with a set
+  of exported names.
+  
+* **Exports**: The exports from a module is a set of names. Exports are always *unqualified* names.
+  (These names can be written with a qualification
+  to disambiguate, if the unqualified name has multiple entries in the environment
+  of the exporting module. But the exported name itself is unqualified.)
+
+* **Imports**: An ``import`` statement names a module and (optionally) a set
+  of *unqualified* names. (This set may be specified via complement, using ``hiding``;
+  this proposal will not worry about this detail.)
+  These names are brought into scope in some fashion:
+
+  * If the import uses ``as``, the module name appearing after the ``as``
+    is known as the module alias. If there is no ``as``, the module name itself
+    is used as the alias.
+
+  * The set of names indicated is brought into scope (that is, put into the
+    global environment), qualified by the module alias in the import.
+
+  * If the import does not use ``qualified``, the names are additionally
+    brought into scope unqualified.
+
+Core Proposal Change Specification
+----------------------------------
+
+This proposal's specification is divided into pieces. This core piece is a necessary
+component of the overall proposal. Later pieces can be chosen piecemeal.
+
+1. A qualified name can now have any number of module qualifications, each
+   separated by a ``.`` and with no intervening whitespace. This can lead
+   to potential ambiguity: is the name ``Data.List.length`` qualified by
+   ``Data.List`` (a well-formed module name) or qualified both by ``Data``
+   and by ``List``? If ``length`` is in scope qualified both by ``Data.List``
+   and by both ``Data`` and ``List``, the occurrence ``Data.List.length`` is
+   an error. There is no way to disambiguate locally without using module
+   aliases or other renamings. (Programmers should seek to avoid this scenario.)
+
+#. Introduce a new extension ``-XLocalModules``.
 
-1. Introduce a new extension ``-XLocalModules``.
+#. Introduce a new concept *export-module*, which is a set of unqualified names
+   and export-modules. Export-modules have names chosen from the module namespace.
+   Instead of exporting a set of unqualified names, modules now export a set of
+   unqualified names and export-modules.
```

Comment:
In other words, modules now export an export-module.

Reply:
I tried writing that at some point, but it failed. It implies that `import M` would actually give access to things like `M.M.x`. I suppose you could modify `import`s to automatically open the outermost export-module, or something, but it seemed easier not go this route.